File: apps/backend/app/schemas/actions.py
"""HTTP DTOs for the recovery action orchestrator."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SchedulerQueueMetrics(BaseModel):
    """Scheduler Queue chips: scheduled, running, delayed, dead-letter."""

    scheduled: int = 0
    running: int = 0
    delayed: int = 0
    dead_letter: int = 0


class ActionDelivery(BaseModel):
    """One sandbox communication attempt."""

    channel: str
    status: str
    provider: str
    provider_message_id: str | None = None
    rate_limited: bool = False
    skipped_reason: str | None = None
    sent_at: datetime | None = None


class ActionExecutionResponse(BaseModel):
    """Execute / schedule / replay result."""

    model_config = ConfigDict(from_attributes=True)

    execution_id: UUID
    recovery_case_id: UUID
    idempotency_key: str
    planner_strategy: str
    action_type: str
    display_status: str
    execution_status: str
    action_chip: str
    scheduled_time: datetime | None = None
    executed_time: datetime | None = None
    retry_attempts: int = 0
    payment_link: str | None = None
    delivery_status: str | None = None
    deliveries: list[ActionDelivery] = Field(default_factory=list)
    request_id: str
    correlation_id: str
    replayed: bool = False
    dead_lettered: bool = False
    policy_reason: str | None = None
    razorpay_resource_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class ActionStatusResponse(BaseModel):
    """Latest execution plus history."""

    recovery_case_id: UUID
    latest: ActionExecutionResponse | None = None
    history: list[ActionExecutionResponse] = Field(default_factory=list)
    active_scheduler_queue: int = 0
    scheduler_queue: SchedulerQueueMetrics = Field(default_factory=SchedulerQueueMetrics)


class ActionDashboardResponse(BaseModel):
    """Merchant-level orchestrator KPIs."""

    scheduled_actions_today: int = 0
    payment_links_sent: int = 0
    successful_retries: int = 0
    failed_deliveries: int = 0
    active_scheduler_queue: int = 0
    scheduler_queue: SchedulerQueueMetrics = Field(default_factory=SchedulerQueueMetrics)
    chips: dict[str, str] = Field(default_factory=dict)

File: apps/frontend/src/types/actions.ts
export const ACTION_CHIPS = ["Scheduled", "Link Sent", "Retrying", "Delivered", "Failed"] as const;

export type ActionChip = (typeof ACTION_CHIPS)[number];

export interface ActionDelivery {
  channel: string;
  status: string;
  provider: string;
  provider_message_id?: string | null;
  rate_limited?: boolean;
  skipped_reason?: string | null;
  sent_at?: string | null;
}

export interface ActionExecution {
  execution_id: string;
  recovery_case_id: string;
  idempotency_key: string;
  planner_strategy: string;
  action_type: string;
  display_status: string;
  execution_status: string;
  action_chip: ActionChip | string;
  scheduled_time?: string | null;
  executed_time?: string | null;
  retry_attempts: number;
  payment_link?: string | null;
  delivery_status?: string | null;
  deliveries: ActionDelivery[];
  request_id: string;
  correlation_id: string;
  replayed: boolean;
  dead_lettered: boolean;
  policy_reason?: string | null;
  razorpay_resource_id?: string | null;
  metadata: Record<string, unknown>;
}

export interface ActionStatus {
  recovery_case_id: string;
  latest: ActionExecution | null;
  history: ActionExecution[];
  active_scheduler_queue: number;
  scheduler_queue?: SchedulerQueueMetrics;
}

export interface SchedulerQueueMetrics {
  scheduled: number;
  running: number;
  delayed: number;
  dead_letter: number;
}

export interface ActionDashboardSummary {
  scheduled_actions_today: number;
  payment_links_sent: number;
  successful_retries: number;
  failed_deliveries: number;
  active_scheduler_queue: number;
  scheduler_queue?: SchedulerQueueMetrics;
  chips: Record<string, string>;
}
